torch/_inductor/cudagraph_trees.py

```diff
@@ -1112,7 +1111,7 @@ def _add_first_outputs(
 
         assert not self.outputs_weakrefs
         for out, static_output_tensor in zip(outputs, self.static_output_tensors):
-            if out is None or static_output_tensor is not None:
+            if not isinstance(out, torch.Tensor) or static_output_tensor is not None:
```

Comment:
Maybe add test case for when we would return dynamic shape output in cudagraph trees ?

I don't think we're hitting the 3rd invocation, reconstruct output path: https://github.com/pytorch/pytorch/blob/072be127d244bd97de34cc4d76264d0a768600d7/torch/_inductor/cudagraph_trees.py#L918-L963

Reply:
Unfortunately, I already have this test case 🤔 

```
    @config.patch({"triton.cudagraphs": True})
    @dynamo_config.patch(
        automatic_dynamic_shapes=True,
        assume_static_by_default=False,
    )               
    def test_dynamic_to_static_cudagraphs(self):
        for b in [False, True]:
            with config.patch({"triton.cudagraph_trees": b}):

                @torch._dynamo.optimize("inductor")
                def fn(x, y):
                    r = x + y
                    return r, r.size(0)

                inputs = (
                    torch.randn((5, 5), device="cuda"),
                    torch.randn((5, 5), device="cuda"),
                )
                self.assertTrue(same(fn(*inputs), (inputs[0] + inputs[1], 5)))
        
                inputs = (
                    torch.randn((6, 6), device="cuda"),
                    torch.randn((6, 6), device="cuda"),
                )
                self.assertTrue(same(fn(*inputs), (inputs[0] + inputs[1], 6)))
```

So I don't understand why this is not exercising these cases...

Reply:
I spent a while trying to understand num_fixed setting in inductor and... I give up. I'm going to argue that the yolov3  is good enough test coverage, and also these changes are "obviously OK" and we can easily keep point fixing these until we've got all the sites.

If you want to argue that we need more clear invariants about the data structures, I'm fine with this, but we should use mypy to work it out, and then apply uniform annotations all throughout cudagraph_trees.py

Reply:
I wonder if it's because we're adding the symints in the front of the calling convention, and that is messing up the num_fixed logic. https://github.com/pytorch/pytorch/blob/main/torch/_functorch/aot_autograd.py#L2990-L2992

I think it's unlikely mypy is going to fix our bug here, although I can add more annotations to cudagraph_trees.

Reply:
@eellison I also thought about this, but it doesn't seem like it should be it. Input SymInt is passed in as example values, so it should not "count" as fixed.